### ISSUE TITLE
(Debugger) Can't BLOAD to Language Card

### ISSUE THREAD

**Issue by Michaelangel007** [AppleWin/AppleWin#256]
BLOAD has been extended to write to aux memory.

Another bank should be supported to write to the Language Card instead of doing this natively via:

```
C081
D000<D000.FFFFM
FA85:4C 69 FF
C080
```

**Comment by Michaelangel007** [comment]
Tom,

The c.s.a2 thread [Memory card for APPLE II](https://groups.google.com/forum/#!topic/comp.sys.apple2/WFOnWGLkl2Q0) has me thinking about debugger support for the Language Card -- Ramworks, Saturn, etc.

Is there a preferred debugger command you would like to use in order to view LC banks 1 and 2 ?  I'm thinking along the lines of:

* `bank 1`
* `bank 2`
* `lc #`

I know you did some work on extended bsave/bload to aux memory and was wondering what your thoughts were on this as well.

**Comment by tomcw** [comment]
Hi Michael,

Yes, some LC support would be good!

I have just captured a year-old request as a GitHub issue here:

https://github.com/AppleWin/AppleWin/issues/406
https://github.com/AppleWin/AppleWin/issu

It'd be good to show the current soft-switch settings.

---

FYI, the issue for extended bsave/bload is here:

https://github.com/AppleWin/AppleWin/issues/166

This extends to support 64K banks as an optional prefix to the address, eg bsave
"dhgr1_aux.bin",1:2000,2000.

Where the command "schema" is:

// Physical 64K memory bank:
// BSAVE [“Filename”] , bank : addr , len
// BSAVE [“Filename”] , bank : addr : end

NB. The addr is physical (not affected by the MMU), so if you save:

* $C000-CFFF then you'll get LC 4K Bank1 (not I/O + firmware space)
* $D000-DFFF then you'll get LC 4K Bank2 (not the D0 ROM)
* $E000-FFFF then you'll get LC 8K (not ROM)

Tom

**Comment by tomcw** [comment]
So given this, using BLOAD you can already load to the LC, eg:
> BLOAD "filename.bin", 0:D000, len

NB. You **must** specific the "0:" bank prefix otherwise `CmdMemorySave (int nArgs)` will just use `mem` pointer instead of calling `MemGetBankPtr(nBank)`.

This is pretty subtle and not an intuitive user experience!

**Comment by Michaelangel007** [comment]
> It'd be good to show the current soft-switch settings.

Agreed!

I've started on this a while back -- it isn't complete -- but I'll try to get some of the cleaned up code checked this weekend.

**Comment by Michaelangel007** [comment]
Related LC support:

#27 
#166 
#256
#406

**Comment by Michaelangel007** [comment]
Yeah having to require "0:" is pretty dumb -- that should be the default argument!

**Comment by tomcw** [comment]
>Yeah having to require "0:" is pretty dumb -- that should be the default argument!

Actually there's a subtle difference:
- without the "0:" then it's the active memory (C000-CFFF is I/O and firmware; D000-FFFF may be ROM or RAM), so reflects the state of the MMU
  - NB. Since these cmds just access via `mem` pointer then C000-C0FF is always 0x00, and C100-CFFF will either return 0x00 or ROM, depending on whether the internal ROM or the slot ROM is mapped in (and if the card even has a ROM)
- with the "0:" then it's the physical RAM

So if you make "0:" the default (so it can be omitted) then there's no way to BSAVE eg. the ROM... or if we chose to support it, there would be no way to BSAVE a Mockingboard's (in slot 4) 6522 registers from C400-C40F.

Maybe this active `mem` isn't important, and we just support the physical memory (then we can make  "0:" the default and it can be omitted)?

**Comment by sicklittlemonkey** [comment]
My 2c. I like the current behaviour of no bank supplied meaning access the
current configuration, and needing a bank supplied to explicitly access
particular RAM banks.

But I'm interested in other ways it could work.

Cheers,
Nick.
